pkg/controller/factory/base_controller.go

```diff
@@ -0,0 +1,158 @@
+package factory
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// baseController represents generic Kubernetes controller boiler-plate
+type baseController struct {
+	name                     string
+	cachesToSync             []cache.InformerSynced
+	sync                     func(ctx context.Context, controllerContext SyncContext) error
+	resyncEvery              time.Duration
+	syncContext              SyncContext
+	shutdownContext          context.Context
+	terminationGraceDuration time.Duration
+}
+
+var _ Controller = &baseController{}
+
+func (c *baseController) Run(ctx context.Context, workers int) {
+	// shutdownContext is used to signal whether a controller has gracefully shutdown
+	shutdownContext, shutdownComplete := context.WithCancel(context.TODO())
+	if c.shutdownContext != nil {
+		panic(fmt.Sprintf("controller %q is already running", c.name))
+	}
+	c.shutdownContext = shutdownContext
+
+	// HandleCrash recovers panics
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting %s ...", c.name)
+	if !cache.WaitForCacheSync(ctx.Done(), c.cachesToSync...) {
+		// this will be recovered using HandleCrash()
+		panic("timeout waiting for informer cache")
+	}
+	klog.V(5).Infof("Caches synced for controller %s", c.name)
+
+	var workerWaitGroup sync.WaitGroup
+
+	// queueContext is used to track and initiate queue shutdown
+	queueContext, queueShutdownComplete := context.WithCancel(context.TODO())
+
+	for i := 1; i <= workers; i++ {
+		klog.Infof("Starting #%d worker of %s controller ...", i, c.name)
+		workerWaitGroup.Add(1)
+		go func() {
+			defer func() {
+				klog.Infof("Shutting down worker of %s controller ...", c.name)
+				workerWaitGroup.Done()
+			}()
+			c.runWorker(queueContext)
+		}()
+	}
+
+	// runPeriodicalResync is independent from queue:w
+	go c.runPeriodicalResync(ctx, c.resyncEvery)
+
+	// Handle controller shutdown
+
+	<-ctx.Done() // wait for controller context to be cancelled
+	klog.Infof("Shutting down %s ...", c.name)
+
+	c.syncContext.Queue().ShutDown() // shutdown the controller queue first
+	for !c.syncContext.Queue().ShuttingDown() {
+		time.Sleep(10 * time.Millisecond) // don't race too much with the queue
+	}
+	queueShutdownComplete() // signal workers the queue is terminated
+
+	// wait for all workers to finish their jobs
+	workerWaitGroup.Wait()
```

Comment:
the purpose of this Wait() is to hang until all workers are finished right? :-) I can move it defer, I just don't like defer because it makes code navigation harder...

Reply:
the same pattern and reasons as Unlock for mutex

in some other language this would be a destructor, so you don't miss calling it on early return or hidden code path.